### ISSUE TITLE
Enhance LC (Rust) by integrating Renegade

### ISSUE THREAD

**Issue by TPhoenixyz** [artificial-intelligence-edge/lorentzian-classification#7]
### Context:
Recently i was watching the latest [Freenet presentation by Ian <@sanity> Clarke](https://www.youtube.com/watch?v=3RBNboYUlVI&t=1438s)
and at 24 minutes he was explaining their "Renegade" algorithm, 
which they use to help connected peers find the desired contracts.

### What is [Renegade](https://github.com/sanity/renegade)?
Renegade is a KNN-based learner, implemented in Rust, with a learned metric,
which means that it predicts future outcomes under certain circumstances,
based on similar circumstances of the past.
At this point you can already tell why i had to think of LC immediately.

### What are the standout features?
- avoids hyperparameters and there is no preprocessing pipeline to configure
- handles mixed numeric and categorical features
- automatically selects K, learns the important features and can drop the unimportant ones
- can take in new datapoints without having to retrain
- more on this is described in its README

### Why is this so interesting for us?
The LC Indicator provides a lot of settings and [requires feature engineering](https://ai-edge.io/docs/indicators/lorentzian-classification/feature-engineering).

> "Conversely, poorly chosen or overly redundant features can lead to overfitting, slow computations, or misleading results."

Renegade was designed to solve two main problems:
1. reliable and efficient future prediction based on past circumstances
2. and doing this without hyperparameters, as @sanity wrote:

> "No hyperparameters — every parameter is an opportunity for misconfiguration"

### How could we go about this?
Now, LC and Renegade try to do similar things, but their approaches are different.
As far as i understand it Lorentzian Distance tries to avoid weighting feature differences equally by dampening outliers,
while Renegades Gower distance, per-feature, figures out which of them matter at all.

The way LC KNN-algorithm selects nearest neighbours may be constrained
due to how the different languages it is implemented in, like PineScript, work.

Since Renegade is implemented in Rust, and LC now has its port in this ecosystem too,
these language constraints are not present here, so Renegades [VP-tree](https://en.wikipedia.org/wiki/Vantage-point_tree) should do a better job here.

Renegade automatically figures out the correct K, instead of it being a user input.

### Ideas for a combined design:
1. Improving outlier-robustness 
   - by keeping Renegade's feature distances, 
   - but applying Lorentzian transform (instead of just a raw normalized diff) 
2. Automate feature selection
   - by throwing more indicators at the algorithm
   - and having it find the feature metrics that actually carry a signal
3. There may be potential to improve LC's forward classification through Renegade's neighbors class voting
4. Renegade could learn whether neighbors are trustworthy, even before considering the indicator signals as rules
5. Use Renegade to retain data by any rules instead of a fixed bar lookback period

### Necessary Disclaimer:
AI was involved in comparing the algorithms to find initial similarities, differences
and a potential combined design, since i'm not an expert in the exact mathematics behind this.

**Comment by jdehorty** [comment]
@TPhoenixyz - This is very interesting. I think it merits its own potential project outside of this one. That would also allow this repo to stay focused purely on identical functionality across ecosystems.

I will put it on the backlog to investigate this further and will keep you updated.

Thanks again for your active contributions and ideas.
